File: solutions/haskell/secret-handshake/1/src/SecretHandshake.hs
module SecretHandshake (handshake) where

import Data.Bits  ( testBit )
import Data.Maybe ( catMaybes )

handshake :: Int -> [String]
handshake n = maybeReverse . catMaybes . map ($ n) $ secrets
  where secrets      = [one, ten, hundred, thousand]

        one          = ifBit 0 (Just "wink")            Nothing
        ten          = ifBit 1 (Just "double blink")    Nothing
        hundred      = ifBit 2 (Just "close your eyes") Nothing
        thousand     = ifBit 3 (Just "jump")            Nothing

        maybeReverse = ifBit 4 reverse id n

        ifBit i a b  = \m -> if testBit m i then a else b

File: solutions/go/secret-handshake/1/secret_handshake.go
package secret
func Handshake(code uint) []string {
	s := []string{}
	if code&uint(1) > 0 {
		s = append(s, "wink")
	}
	if code&uint(2) > 0 {
		s = append(s, "double blink")
	}
	if code&uint(4) > 0 {
		s = append(s, "close your eyes")
	}
	if code&uint(8) > 0 {
		s = append(s, "jump")
	}
	if code&uint(16) > 0 {
		for i, j := 0, len(s)-1; i < j; i, j = i+1, j-1 {
			s[i], s[j] = s[j], s[i]
		}
	}
	return s
}